R/data.table.R

```diff
@@ -1861,90 +1861,243 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
+  # otherwise non-atomic types (e.g. S4). Further, we want to minimize
+  # the number of checks done, e.g. if all columns are already the same type
+
+  col.classes = lapply(X, class)
+  uniq.col.classes = unique(col.classes) # list structure preserves cases where a column has multiple classes
+  uniq.classes = unique(unlist(uniq.col.classes)) # vector of distinct classes used to determine checks to proceed
+  
+  # Check for ff columns which need to be brought into memory. Running the any
+  # command on the unique column class list means for x with many columns we 
+  # don't do expensive per-column checks unless necessary.
+  any.ff = ("ff" %chin% uniq.classes)
+  if (any.ff) {
+    which.ff = which(sapply(col.classes, function(cl_vec) { "ff" %chin% cl_vec }))
+    for (j in which.ff) {
+      X[[j]] = X[[j]][] # bring into memory
+    }
+    
+    # need to redetermine the classes now that all columns have been brought into memory
+    col.classes = lapply(X, class)
+    uniq.col.classes = unique(col.classes)
+    uniq.classes = unique(unlist(uniq.col.classes))
+  }
+  
+  # Next determine if any columns are list or non-atomic type. If so, convert all columns to lists
+  col.types = sapply(X, typeof)
+  uniq.col.types = unique(col.types)
+  atomics <- c("logical", "integer", "numeric", "double", "complex", "character", "raw")
+  any.non.atomic = (length(setdiff(uniq.col.types, atomics)) > 0L)
+  if (any.non.atomic) {
     for (j in seq_len(p)) {
-      xj = X[[j]]
-      if (is.recursive(xj)) { }
-      else X[[j]] = as.list(as.vector(xj))
+      if (!is.recursive(X[[j]])) { 
+        X[[j]] = as.list(as.vector(X[[j]]))
+      }
     }
-  }
-  else if (all.logical) { }
-  else if (non.numeric) {
-    for (j in seq_len(p)) {
-      if (is.character(X[[j]])) next
-      xj = X[[j]]
-      miss = is.na(xj)
-      xj = if (length(levels(xj))) as.vector(xj) else format(xj)
-      is.na(xj) = miss
-      X[[j]] = xj
+    
+    # redetermine classes and types for next checks
+    col.classes = lapply(X, class)
+    uniq.col.classes = unique(col.classes)
+    uniq.classes = unique(unlist(uniq.col.classes))
+    col.types = sapply(X, typeof)
+    uniq.col.types = unique(col.types)
+  }
+  
+  # Convert factors to character vectors
+  any.factors = ("factor" %chin% uniq.classes)
+  if (any.factors) {
+    which.factors = which(sapply(col.classes, function(cl_vec) { "factor" %chin% cl_vec }))
+    for (j in which.factors) {
+      miss = is.na(X[[j]])
+      X[[j]] = as.vector(X[[j]])
+      is.na(X[[j]]) = miss
+    }
+    
+    # redetermine classes and types for next checks
+    col.classes = lapply(X, class)
+    uniq.col.classes = unique(col.classes)
+    uniq.classes = unique(unlist(uniq.col.classes))
+    col.types = sapply(X, typeof)
+    uniq.col.types = unique(col.types)
+  }
+  
+  # Classes to be converted to character vectors
+  charconvert.classes = c("Date", "POSIXct", "POSIXlt", "raw")
```

Comment:
this looks the same as the `factor` case, should it be done together?

Reply:
`factor` is slightly different, in the original as.matrix.data.table code it was converted with `as.vector()`, where `Date`, `POSIXct`, and `POSIXlt` were converted with `format()`. Instead of two similar code blocks we could simplify by putting an if statement in the for loop that handles the factor case.